### PR TITLE
Fix rust linting with cargo when multiple targets are present

### LINKED ISSUE #302: Cargo doesn't run when there is a bin and a lib
`cargo rustc` can only run when there is only one target so when there is multiple targets, there is no linting with cargo
Here is the cargo error: 
```
cargo rustc -- --error-format=json -Z no-trans
error: extra arguments to `rustc` can only be passed to one target, consider filtering
the package by passing e.g. `--lib` or `--bin NAME` to specify a single target
```

I think you should use `cargo build --message-format json`

### PR DESCRIPTION
**Pull request by notkild**
Fix #302

### CODE REVIEW

File: ale_linters/rust/cargo.vim [hunk 1]
```diff
@@ -3,19 +3,89 @@
...
 function! ale_linters#rust#cargo#GetCargoExecutable(bufnr) abort
-    if ale#util#FindNearestFile(a:bufnr, 'Cargo.toml') !=# ''
-        return 'cargo'
-    else
-        " if there is no Cargo.toml file, we don't use cargo even if it exists,
-        " so we return '', because executable('') apparently always fails
-        return ''
-    endif
+	if ale#util#FindNearestFile(a:bufnr, 'Cargo.toml') !=# ''
+		return 'cargo'
```
Comment (w0rp) [inline ale_linters/rust/cargo.vim]: Could you change your tab indents to four spaces, so it matches the rest of the codebase?

Comment (w0rp) [inline ale_linters/rust/cargo.vim]: It looks like these changes are just from re-indenting everything.

File: ale_linters/rust/cargo.vim [hunk 2]
```diff
@@ -3,19 +3,89 @@
...
+	else
+		" if there is no Cargo.toml file, we don't use cargo even if it exists,
+		" so we return '', because executable('') apparently always fails
+		return ''
+	endif
+endfunction
+
+function! ale_linters#rust#cargo#HandleCargoErrors(buffer_number, errorlines) abort
```
Comment (w0rp) [inline ale_linters/rust/cargo.vim]: The indentation here makes this hard to read. How does this differ from the rustc handler?

Comment (notkild) [inline ale_linters/rust/cargo.vim]: Yup sorry I will do that

Comment (notkild) [inline ale_linters/rust/cargo.vim]: And error message from cargo build are different, JSON error is contained in a 'message' object so I have to copy function from rustc to fix differences

**Comment by notkild** [comment]
Is this good now?

File: ale_linters/rust/cargo.vim (hunk 2 above)
Comment (notkild) [inline ale_linters/rust/cargo.vim]: Here is output from `cargo rustc --lib -- --error-format=json -Z no-trans`:
```
{"message":"expected one of `!` or `::`, found `hyper`","code":null,"level":"error","spans":[{"file_name":"src/crawl/mod.rs","byte_start":1529,"byte_end":1534,"line_start":7,"line_end":7,"column_start":4,"column_end":9,"is_primary":true,"text":[{"text":"se hyper::Url;","highlight_start":4,"highlight_end":9}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
```

And here is the output from `cargo build --message-format json`:
```
{"message":{"children":[],"code":null,"level":"error","message":"expected one of `!` or `::`, found `hyper`","rendered":null,"spans":[{"byte_end":1534,"byte_start":1529,"column_end":9,"column_start":4,"expansion":null,"file_name":"src/crawl/mod.rs","is_primary":true,"label":null,"line_end":7,"line_start":7,"suggested_replacement":null,"text":[{"highlight_end":9,"highlight_start":4,"text":"se hyper::Url;"}]}]},"package_id":"unibot 0.1.0 (path+file:///home/notkild/projects/rust/unibot)","reason":"compiler-message","target":{"kind":["lib"],"name":"libunibot","src_path":"/home/notkild/projects/rust/unibot/src/lib.rs"}}
```

**Comment by w0rp** [comment]
Isn't this error handling function almost exactly the same as the function for `rustc`? Can't we fix the existing handler so it can work with both? It looks like you could set `l:error` to `l:error.message` if such a key exists, and then the rest is the same.

**Comment by notkild** [comment]
I think there is a way but I am pretty bad at vimscript but I will try to find another way to do that

**Comment by notkild** [comment]
Ok so I think I have found a solution. What do you think about that?

**Comment by w0rp** [comment]
Looks good to me. :+1: I'll merge this, and let Rust users give it a go.

**Comment by notkild** [comment]
Yup, if there any issue, send me a mail :smile:

**Comment by vimeitor** [comment]
This commit breaks linting for me.

`
Error detected while processing function <SNR>98_HandleExitVim[1]..<SNR>98_HandleExit[27]..ale_linters#rust#rustc#HandleRustcErrors:
`

`
E121: Undefined variable: code
E116: Invalid arguments for function empty(l:error.code) && index(g:ale_rust_ignore_error_codes, l:error.code.code) > -1
E15: Invalid expression: !empty(l:error.code) && index(g:ale_rust_ignore_error_codes, l:error.code.code) > -1
`

**Comment by notkild** [comment]
Are you using rustc to lint code?

**Comment by vimeitor** [comment]
I haven't touched the default config.

**Comment by w0rp** [comment]
See issue #256. I just pushed a commit which might fix this now.

**Comment by vimeitor** [comment]
Almost :)

`
Error detected while processing function <SNR>100_HandleExitVim[1]..<SNR>100_HandleExit[27]..ale_linters#rust#rustc#HandleRustcErrors:
E715: Dictionary required
`

**Comment by w0rp** [comment]
Try again one more time...

**Comment by vimeitor** [comment]
Seems to work wonderfully now!

Thank you for the work both of you =)

**Comment by w0rp** [comment]
Cool!

It was just down to a bug in a condition above not checking if the output looks like Cargo output or `rustc` output correctly. I juggled things around a bit until it started working again.

**Comment by w0rp** [comment]
@notkild BTW, if you have some example output from `cargo` I could use for testing, that would be useful. We could easily write some Vader tests to cover the function for handling Rust errors.

**Comment by notkild** [comment]
Yup if you want, I will make some different error to see different output

**Comment by notkild** [comment]
Here are some examples of error and warning.

[error1.txt](https://github.com/w0rp/ale/files/758140/error1.txt)
[error2.txt](https://github.com/w0rp/ale/files/758143/error2.txt)
[error3.txt](https://github.com/w0rp/ale/files/758141/error3.txt)
[warning1.txt](https://github.com/w0rp/ale/files/758142/warning1.txt)

**Comment by w0rp** [comment]
Thanks for the output! I missed that. I created some Vader tests now with some output I took from my machine. I might add those into test cases later.

**Comment by notkild** [comment]
Good idea, I can give you some output when cargo compile `feature`, it give us JSON output that is different than error or warning

**Pull request merged**